osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs

```diff
@@ -142,23 +149,32 @@ public async Task JoinRoom(Room room)
         /// <returns>The joined <see cref="MultiplayerRoom"/>.</returns>
         protected abstract Task<MultiplayerRoom> JoinRoom(long roomId);
 
-        public virtual Task LeaveRoom()
+        public Task LeaveRoom()
         {
-            Scheduler.Add(() =>
-            {
-                if (Room == null)
-                    return;
+            if (Room == null)
+                return Task.FromCanceled(new CancellationToken(true));
```

Comment:
I'm not quite experienced with this so this may be a silly comment, but the fact that this null room check and the reset room schedule below are outside of the task chain means there's still a potential for the issue to occur, which is if `LeaveRoom` is invoked while in the middle of executing `JoinRoom`, before `Room` is set to non-null value.

I've [injected both `JoinRoom` and `LeaveRoom` with logs](https://gist.github.com/frenzibyte/d9f45c00736ad10afda3fca307771dc8) and that's what I got when I very quickly left a room after joining it, and the output matches what I was saying above:
```
2021-01-26 16:51:27 [verbose]: [JoinRoom] Beginning join process
2021-01-26 16:51:27 [verbose]: [LeaveRoom] Beginning leaving process
2021-01-26 16:51:27 [verbose]: [LeaveRoom] Room is null, returning...
2021-01-26 16:51:27 [verbose]: [JoinRoom] Joined room server-side
2021-01-26 16:51:27 [verbose]: [JoinRoom] Users populated
2021-01-26 16:51:27 [verbose]: [JoinRoom] New room assigned
2021-01-26 16:51:27 [verbose]: [JoinRoom] Local room settings updated
```

Notice `LeaveRoom` early-returned due to `Room == null`, as it isn't set on `JoinRoom` yet.

The obvious fix would be to move all of this inside the chain, but I still don't get what the attached comment below is referring to exactly, is it about the user will transition to an invalid room screen when pressing "create button" because of `Room != null`? because from the logs I attached above, `JoinRoom` will still set `Room` to a value, causing the create button issue to still persist.

Reply:
[you've rediscovered the issue i already pointed out](https://github.com/ppy/osu/pull/11551#discussion_r563969231). there don't seem to be any new changes since that would resolve it.

Reply:
Yeah essentially it is the same issue, but I thought that conversation was considered not possible to occur, so I posted a new one writing about it in detail of how it could happen, wasn't sure which is best but at least issue is pointed out.